rfcs/20210301-computation-datatype/README.md

```diff
@@ -0,0 +1,178 @@
+typically require quantization parameters that depend on the
+computation data-type. For floating point data-types
+- `bf16` and `f32` are compatible for both training and inference,
+- and `f16` and `f32` only for inference since `f32` and `f16` are not
+  typically interchangeable for training.
+
+Proposal: support only down-conversion to "compatible"
+data-types. This will allow to restrict the feature to only commonly
+used combinations and would make internal logic and validation
+simpler. Here we would return `invalid_argument` if users pass
+incompatible data-type for down-conversion hint/requirement, to help
+them debug incorrect logic.
+
+## Down-conversion policy: requirement vs hint.
+Here the question is if the user specifies a target type to
+down-convert to, should we treat it as a requirement (we can
+down-convert only to this type) or should it be a hint (we can
+down-convert to a type other than the one specified as long as it is
+"compatible").
+
+Proposal: support only hints, to allow library for optimal dispatching
+without user needing to tweak down-conversion policy for better
+performance. Also, this is more scalable as we introduce new
+data-types (e.g. if we were to introduce `tf32`).
+
+## Per tensor vs per primitive setting
+This would be the granularity at which the user would allow conversion
+to happen. Per tensor seems to have the most flexibility, however it
+is unlikely that users would need this level of flexibility. For the
+per-primitive setting, the implicit conversion will be assumed to
+happen only for input tensors, and not output.
+
+Proposal: support only per primitive setting. The main points against
+per tensor are: a) it would require more user changes to enable this
+feature and b) the implicit down-convert feature is planned to be
+exposed to end-user (FWK or toolkit user), and there is no plan to
+expose any fine grain control there (it will be global or scoped to a
+sub-graph).
+
+## Allow down-conversion through reorder vs hide it in execution
+
+This is basically deciding what data-type should be returned when a
+primitive is queried for one of its memory descriptor when implicit
+down-conversion happens. One option is to preserve the data-type
+passed by user so that down-conversion is fully hidden and
+implicit. The other option is to return a memory descriptor with the
+data-type that will be used during computation. The first option has
+the benefit to completely hide the down-conversion and does not
+require the user to add logic to handle potential down-conversion
+through reorder if creation data-type is different than queried
+data-type. The second option has the benefit of allowing to
+down-convert ahead of time and reduce the execution time (e.g. weights
+down-conversion can be computed ahead of time during inference).
+
+Proposal: expose the data-type we down-convert to only when the user
+data layout is `any`. When the user does not pass `any`, we cannot
```

Comment:
Since `any` is currently used to work with blocked formats, can this approach become a problem if users will switch to using plain formats?

Reply:
When user switch to plain layout, they would no more pass `any` for that given memory but the plain layout they want.
And under those circumstances, oneDNN should return the same memory descriptor as the one the user passed upon query,